### PR TITLE
Forward opacity and blend mode down the tree

### PR DESCRIPTION
**Pull request by jonahwilliams**
This allows the elimination of simple opacity and blend modes in limited circumstances. TBD question about stacked opacity.

### CODE REVIEW

File: packages/vector_graphics_compiler/test/opacity_peephole_test.dart [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    final ResolvedTextNode textNode =
+        queryChildren<ResolvedTextNode>(newNode).single;
+
+    expect(textNode.paint.fill?.color, const Color(0xffffffff));
+  });
+
+  test('Collapses nested opacity groups', () async {
+    final Node node = await parseAndResolve('''
+<svg viewBox="0 0 200 200">
+  <g opacity="0.5">
```
Comment (jonahwilliams) [inline packages/vector_graphics_compiler/test/opacity_peephole_test.dart:136]: Not actually sure what the right result is here

File: packages/vector_graphics_compiler/lib/src/paint.dart [hunk 2]
```diff
@@ -396,16 +396,16 @@ class Paint {
...
     this.fill,
-  });
+  }) : blendMode = blendMode ?? BlendMode.srcOver;
 
   /// The Porter-Duff algorithm to use when compositing this painting object
   /// with any objects painted under it.
   ///
   /// Defaults to [BlendMode.srcOver].
-  final BlendMode? blendMode;
+  final BlendMode blendMode;
```
Comment (jonahwilliams) [inline packages/vector_graphics_compiler/lib/src/paint.dart:408]: Made the default actually srcOver. This required some test updates

File: packages/vector_graphics_compiler/test/end_to_end_test.dart [hunk 3]
```diff
@@ -86,11 +86,11 @@ void main() {
         offsets: Float32List.fromList(<double>[0, 1]),
         tileMode: 0,
       ),
-      const OnPaintObject(
+      OnPaintObject(
         color: 0xffffffff,
         strokeCap: 1,
         strokeJoin: 1,
-        blendMode: 0,
+        blendMode: BlendMode.srcOver.index,
```
Comment (jonahwilliams) [inline packages/vector_graphics_compiler/test/end_to_end_test.dart:93]: These changes are related to changing the BlendMode default value

File: packages/vector_graphics_compiler/test/parser_test.dart [hunk 4]
```diff
@@ -174,14 +174,11 @@ void main() {
 
     expect(instructions.paints, const <Paint>[
```
Comment (jonahwilliams) [inline packages/vector_graphics_compiler/test/parser_test.dart:175]: See? it works! free blend mode

File: packages/vector_graphics_compiler/test/opacity_peephole_test.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/test/opacity_peephole_test.dart:136]: multiplicative

File: packages/vector_graphics_compiler/test/opacity_peephole_test.dart [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    </g>
+  </g>
+</svg>''');
+
+    final OpacityPeepholeOptimizer visitor = OpacityPeepholeOptimizer();
+    final Node newNode = visitor.apply(node);
+    final ResolvedPathNode textNode =
+        queryChildren<ResolvedPathNode>(newNode).single;
+
+    expect(textNode.paint.fill?.color, const Color(0x07ffffff));
```
Comment (dnfield) [inline packages/vector_graphics_compiler/test/opacity_peephole_test.dart:150]: Need to update this.

File: packages/vector_graphics_compiler/lib/src/geometry/basic_types.dart [hunk 6]
```diff
@@ -104,6 +104,18 @@ class Rect {
...
+  /// Whether or not the rect has any contents.
+  bool get isEmpty => width == 0 || height == 0;
+
+  /// Whether or not [other] intersect this rectangle.
+  bool intersects(Rect other) {
+    if (isEmpty || other.isEmpty) {
+      return false;
+    }
+    return math.max(left, other.left) < math.min(right, other.right) &&
+        math.max(top, other.top) < math.min(bottom, other.bottom);
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/geometry/basic_types.dart:116]: Consider adding a comment that this method only works for sorted rects. We may also want some kind of method to sort rects. So far it hasn't been a problem but at some point we'll get a rect where this.bottom < this.top for example.

Alternatively, we just assert in the ctor that it's sorted...

File: packages/vector_graphics_compiler/lib/src/draw_command_builder.dart [hunk 7]
```diff
@@ -14,8 +14,6 @@ class DrawCommandBuilder {
 
   /// Add a save layer to the command stack.
   void addSaveLayer(Paint paint) {
-    assert(paint.fill!.color != null);
```
Comment (dnfield) [inline packages/vector_graphics_compiler/lib/src/draw_command_builder.dart:17]: Calling saveLayer without a color on it is broken at runtime IIRC. Why are we removing this assert?

At the very least I think we'd need opaque black on here - is that added elsewhere?

Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/draw_command_builder.dart:17]: Fill colors default to opaque black and Fill.color is no longer nullable

File: packages/vector_graphics_compiler/lib/src/geometry/basic_types.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/geometry/basic_types.dart:120]: I added an assert and comment here for now

File: packages/vector_graphics_compiler/test/opacity_peephole_test.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/test/opacity_peephole_test.dart:150]: ?

File: packages/vector_graphics_compiler/lib/src/paint.dart [hunk 8]
```diff
@@ -593,10 +589,9 @@ class Fill {
   String toString() {
     final StringBuffer buffer = StringBuffer('Fill(');
     String leading = '';
-    if (color != null) {
-      buffer.write('${leading}color: $color');
-      leading = ', ';
-    }
+    buffer.write('${leading}color: $color');
+    leading = ', ';
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/paint.dart]: done

File: packages/vector_graphics_compiler/lib/src/paint.dart [hunk 9]
```diff
@@ -454,10 +454,8 @@ class Paint {
   String toString() {
     final StringBuffer buffer = StringBuffer('Paint(');
     String leading = '';
-    if (blendMode != null) {
-      buffer.write('${leading}blendMode: $blendMode');
-      leading = ', ';
-    }
+    buffer.write('${leading}blendMode: $blendMode');
+    leading = ', ';
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/vector_graphics_compiler/lib/src/paint.dart]: done

File: packages/vector_graphics_compiler/lib/src/draw_command_builder.dart (hunk 7 above)
Comment (dnfield) [reply packages/vector_graphics_compiler/lib/src/draw_command_builder.dart:17]: Ahhh ok.

File: packages/vector_graphics_compiler/test/opacity_peephole_test.dart (hunk 5 above)
Comment (dnfield) [reply packages/vector_graphics_compiler/test/opacity_peephole_test.dart:150]: I misread this as 0x7f instead of 0x07f

**Review by dnfield — APPROVED** [review]

**Comment by jonahwilliams** [comment]
wooo! See, its not that bad :)

**Pull request merged**